lib/tree.rb

```diff
@@ -16,51 +16,196 @@ def initialize
+  def find_helper(current, key)
+    return nil if current.nil? 
+
+    if current.key == key 
+      return current.value
+    elsif key < current.key 
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # iterative find helper
+  # def find_helper(current, key)
+  #   return nil if current.nil? 
+  #   found = false
+  #   until found 
+  #     if current.key == key 
+  #       found = true 
+  #       return current.value 
+  #     else
+  #       if key < current.key 
+  #         current = current.left 
+  #       else
+  #         current = current.right 
+  #       end
+  #     end
+  #     return nil if current.nil? 
+  #   end
+  # end
+
+  # Time Complexity: O(n) to visit all nodes
+  # Space Complexity: O(n) values array depends on how many nodes there are 
   def inorder
-    raise NotImplementedError
+    values = []
+    return inorder_helper(@root, values)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, values)
+    return values if current.nil?
+
+    inorder_helper(current.left, values)
+    values.push({key: current.key, value: current.value})
+    inorder_helper(current.right, values)
+    return values 
+  end
+
+  # Time Complexity: O(n) to visit all nodes
+  # Space Complexity: O(n) values array depends on how many nodes there are 
   def preorder
-    raise NotImplementedError
+    values = []
+    return preorder_helper(@root, values)
+  end
+
+  def preorder_helper(current, values)
+    return values if current.nil? 
+
+    values.push({key: current.key, value: current.value})
+    preorder_helper(current.left, values)
+    preorder_helper(current.right, values)
+    return values
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) to visit all nodes
+  # Space Complexity: O(n) values array depends on how many nodes there are 
   def postorder
-    raise NotImplementedError
+    values = []
+    return postorder_helper(@root, values)
+  end
+
+  def postorder_helper(current, values)
+    return values if current.nil? 
+
+    postorder_helper(current.left, values)
+    postorder_helper(current.right, values)
+    values.push({key: current.key, value: current.value})
+    return values
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) to visit all the nodes 
+  # Space Complexity: O(log n) if balanced, O(n) if unbalanced
   def height
-    raise NotImplementedError
+    height_helper(@root)
+  end
+
+  def height_helper(current)
+    return 0 if current.nil? 
+
+    left_height = height_helper(current.left)
+    right_height = height_helper(current.right)
+
+    return left_height > right_height ? left_height + 1 : right_height + 1 
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n^2) due to use of .shift 
+  # Space Complexity: O(n)
+  # Solution from class
   def bfs
```

Comment:
👍 This should work, and good insight with the `.shift`'s affect on the time complexity.